src/sfloader/fluid_defsfont.c

```diff
@@ -3360,38 +3367,52 @@ fixup_sample (SFData * sf)
   int invalid_loops=FALSE;
   int invalid_loopstart;
   int invalid_loopend, loopend_end_mismatch;
-  unsigned int total_samples = sf->samplesize / FLUID_MEMBER_SIZE(fluid_defsfont_t, sampledata[0]);
+  unsigned int total_bytes = sf->samplesize;
+  unsigned int total_samples = total_bytes / FLUID_MEMBER_SIZE(fluid_defsfont_t, sampledata[0]);
 
   p = sf->sample;
   while (p)
     {
+      /* Standard SoundFont files (SF2) use sample word indices for sample start and end pointers,
+       * but SF3 files with Ogg Vorbis compression use byte indices for start and end. */
+      unsigned int max_end = (sam->sampletype & FLUID_SAMPLETYPE_OGG_VORBIS) ? total_bytes : total_samples;
+
       sam = (SFSample *) (p->data);
+
+      /* ROM samples are unusable for us by definition, so simply ignore them. */
+      if (sam->sampletype & FLUID_SAMPLETYPE_ROM)
+      {
+        sam->start = sam->end = sam->loopstart = sam->loopend = 0;
+        p = fluid_list_next (p);
+        continue;
+      }
+
+      /* If end is over the sample data chunk or sam start is greater than 4
+       * less than the end (at least 4 samples).
+       *
+       * FIXME: where does this number 4 come from? And do we need a different number for SF3 files?
+       * Maybe we should check for the minimum Ogg Vorbis headers size? */
```

Comment:
>I think it's in sect. 6.3:

Not sure... for looped samples, that section defines a minimum sample length, yes. But from my understanding, it would be 8, not 4. And it doesn't say anything about unlooped samples.
Also, in fluid_sample_import_sfont we do another check on the sample length, this time checking it's at least 8 samples.

> Technically this logic should be valid for SF3 as well. But due to the byte indices, we cannot use 4 here, maybe just check (sam->start >= sam->end) then?

It's not really about the byte indices, but about the fact that we're still dealing with compressed audio here. Checking sample length only makes sense once we have uncompressed the data. But the check for >=4 (or 8) will still work, as a valid Ogg Vorbis is at least 27 bytes in length (the vorbis header).

Hm, so should I remove the FIXME? I'm still not convinced we know where the number 4 comes from :-)

Reply:
The 4 might indeed not be correct (sect. 7.10):

>   Also, to allow a variety of hardware platforms to be able to reproduce the data, the samples have a minimum length of 48 data points, a minimum loop size of 32 data points and a minimum of 8 valid points prior to dwStartloop and after dwEndloop. Thus dwStart must be less than dwStartloop-7, dwStartloop must be less than dwEndloop-31, and dwEndloop must be less than dwEnd-7.  If these constraints are not met, the sound may optionally not be played if the hardware cannot support artifact-free playback for the parameters given. 

Not sure if it's a good idea to enforce this requirement, might unnecessarily break soundfonts. I also havent tested whether fluidsynth can handle such short samples "artifact-free". Maybe just remove one of the sample length checks then? And is this `fluid_sample_t::valid` member even enforced anywhere??

Reply:
> If these constraints are not met, the sound may optionally not be played if the hardware cannot support artifact-free playback for the parameters given.

I think this "optionally" is the key here. It's not really a strict constraint which should lead to a SoundFont being rejected by the player. It's basically just a suggestion for SoundFont editors / designers. And there's a comment in the ChangeLog from Element:
```
2003-06-28  Josh Green  <jgreen@users.sourceforge.net>

	* src/fluid_defsfont.c: Moved call to fluid_voice_optimize_sample
	from fluid_inst_zone_import_sfont to fluid_defsfont_load. Also
	reduced minimum sample size before rejection from 48 to 8 (could
	be lower?).
```
The "could be lower" is interesting, as FluidSynth has worked fine with that 4 sample limit for years. And good point about if sample->valid is being used! Turns out it's only used to prevent fluid_voice_optimize_sample from calculating sample->amplitude_that_reaches_noise_floor. It doesn't affect if that sample is rendered or not. And there isn't anything in that optimization function that depends on a specific length of the sample. Actually, it only depends on loopstart and loopend being valid indices into the sample data array, not even their relative order is important.

As fixup_sample now makes sure that ROM samples have all sample pointers set to 0, this sample->valid flag could be removed. I'll add that to the refactor list and simply remove the FIXME comment from this PR, ok?

Reply:
Alright.

Reply:
Actually, can we leave the FIXME comment in for now and merge this as is? I'm going to refactor that code anyway and keeping it in prevents me from having to rebase by refactor branch again...